PROTOCOL.md

```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
+
+`GroupKeyResponse`s are always encrypted with (`encryptionType 1 (RSA)`) for the `rsaPublicKey` provided in the `GroupKeyRequest`. The `content` encoded as `contentType 0` (JSON) is as follows: encoded as `contentType 0` (JSON) is as follows:
+
+```
+[requestId, streamId, [groupKeys]]
+```
+
+ Field        | Type      | Description
+ ------------ | --------- | -----------
+ `requestId`  | `string`  | The `requestId` of the `GroupKeyRequest`, identifying which request is being responded to.
+ `streamId`   | `string`  | The stream for which a key is being delivered.
+ `groupKeys`  | `array`   | Array of `[groupKeyId, encryptedGroupKey]` pairs, containing the requested keys encrypted with RSA for the public key provided in the `GroupKeyRequest`. The `encryptedGroupKey` is a hex-encoded binary string.
+ 
+Example of a `GroupKeyResponse` including the rest of the `Stream Layer` fields:
+
+```
+[32, [...msgIdFields], [...msgRefFields], 29, 0, 0, ["requestId", "streamId", ["keyId","123abc"]], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyReset
+
+Sent by publishers to each subscribers' key exchange streams whenever they want to re-key the stream (in order to revoke access from some subscribers). 
+
+The requestor should handle this message by adding the contained keys to their key storage.
+
+`GroupKeyReset`s are encrypted with (`encryptionType 1 (RSA)`) and are sent to subscribers whose `rsaPublicKey` is known by the publisher (due to receiving a `GroupKeyRequest` earlier, for example. The `content` encoded as `contentType 0` (JSON) is as follows:
+
+```
+[streamId, groupKeyId, groupKey]
+```
+
+ Field        | Type      | Description
+ ------------ | --------- | -----------
+ `streamId`   | `string`  | The stream for which a key is being delivered.
+ `groupKeyId` | `string`  | The id of the `groupKey`.
+ `groupKey`   | `string`  | The group key as a hex-encoded binary string.
+ 
+Example of a `GroupKeyReset` including the rest of the `Stream Layer` fields:
+
+```
+[32, [...msgIdFields], [...msgRefFields], 30, 0, 0, ["keyId", "123abc"], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyErrorResponse
+
+In case there's an error processing a `GroupKeyRequest`, this message gets sent by a publisher to the requestor's key exchange stream. 
+
+`GroupKeyErrorResponse`s are always unencrypted (`encryptionType 0`). The `content` encoded as `contentType 0` (JSON) is as follows:
+
+```
+[requestId, streamId, errorCode, errorMessage]
+```
+
+ Field        | Type      | Description
+ ------------ | --------- | -----------
+ `requestId`  | `string`  | The `requestId` of the `GroupKeyRequest`, identifying which request is being responded to.
+ `streamId`   | `string`  | The stream for which a key is being delivered.
```

Comment:
perhaps should include the `groupKeyId` in the error, clients can figure this out by looking at which request corresponds to which groupKeyId, but might help debugging.

Reply:
The request can have multiple `groupKeyIds`. It would make sense to include in the error response the ones that errored. I'll make that change.

This actually leads us to a situation where _some_ of the requested ids may succeed and some may fail. Perhaps in this case the publisher should send both a `GroupKeyResponse` and `GroupKeyErrorResponse`. I guess that's fine?!

Reply:
Added some text about that, and also considered the case where the publisher has many instances connected to the network and subscribed to the key exchange stream. In this case the requestor will receive many responses for the same request.